### DISCUSSION TITLE
How to cache a function's output?

### DISCUSSION THREAD

**Discussion by muuvmuuv** [marlonrichert/zsh-snap#139] [General]
I would like to cache a function's output, but what I get is "command not found: <function_output_here>". What am I missing?

```zsh
function ca_pw {
  echo $(docker run --rm -v step:/home/step smallstep/step-ca cat secrets/password)
}
znap eval _ca_pw "ca_pw"
```

**Comment by marlonrichert** [comment]
This looks problematic:
```zsh
  echo $(docker run --rm -v step:/home/step smallstep/step-ca cat secrets/password)
```
1. Using `$( ... )` instead of `"$( ... )"` turns all whitespace into word separators, *which changes the meaning of line breaks.*
2. If you use `echo` instead of `print -r -- `, then [almost anything starting with a `\`](https://zsh.sourceforge.io/Doc/Release/Shell-Builtin-Commands.html#index-echo) will be processed as an escape sequence.
3. Using `echo $( <command> )` is completely superfluous. You can just use `<command>` instead (unless you actually want the above).

So, I would rewrite that line like this:
```zsh
docker run --rm -v step:/home/step smallstep/step-ca cat secrets/password
```

Is there a particular reason why don't just do this?
```zsh
znap eval _ca_pw 'docker run --rm -v step:/home/step smallstep/step-ca cat secrets/password'
```

**Reply by muuvmuuv** [reply]
I already tried the last one, which also results in the same error.

```
/Users/marvin/.cache/zsh-snap/eval/_ca_pw.zsh:2: command not found: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

```
; cat /Users/marvin/.cache/zsh-snap/eval/_ca_pw.zsh
#docker\ run\ --rm\ -v\ step:/home/step\ smallstep/step-ca\ cat\ secrets/password
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

**Reply by muuvmuuv** [reply]
That code highlighting tho xD

**Reply by marlonrichert** [reply]
Well, your command `docker run --rm -v step:/home/step smallstep/step-ca cat secrets/password'` apparently outputs `xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx`. That’s not Znap’s fault.

**Reply by muuvmuuv** [reply]
But this is valid, I want to cache the return value somewhere. It does not change but is stored in the container.

**Reply by marlonrichert** [reply]
What do you mean with “the return value”? Shell commands don’t have return values; they only have a standard output stream, a standard error stream and an exit status. `znap eval` caches the standard output.

**Reply by muuvmuuv** [reply]
I mean the standard output sorry for the wording. Usually, I need to run the docker command each day when certificates need to be updated (the standard output of the command is the password for the CA Server), since the password does not change but is stored in the container I would like to cache it so the command does not always take several seconds to run.

**Reply by marlonrichert** [reply]
> the standard output of the command is the password for the CA Server

A password is not a command, is it? You cannot `eval` a password. So, don't use `znap eval` for that. Just redirect the output to a file. Or alternative, change your function to output code that does something with the password, instead of just outputting the password. Then you can use it with `znap eval`.

**Reply by muuvmuuv** [reply]
With command, I mean the docker command, not the password. I did expect that znap eval caches the standard output, so I just misunderstood that point, I guess? Redirecting to a file would work for now, just thought I could use znap for it.

**Reply by marlonrichert** [reply]
I could separate the caching part of `znap eval` into a separate command, but the problem here is: How would the code know when to invalidate the cache? For this case, I don’t think it can. You would end up having to do it manually anyway.

**Reply by muuvmuuv** [reply]
That sounds great, but I don't think this is worth your time if its just for my specific case. 

Anyway, invalidating manually is fine in my case, I would only need to do that one time per month when I reset my containers (so the password will change).
